skimage/filters/thresholding.py

```diff
@@ -490,3 +490,111 @@ def find_local_maxima(hist):
             return bin_centers[upper_bound]
         elif bias == 'mid':
             return bin_centers[(threshold + upper_bound) // 2]
+
+
+def threshold_mean(image):
+    """Return threshold value based on the mean of grayscale values.
+
+    Parameters
+    ----------
+    image : (N, M) ndarray
+        Grayscale input image.
+
+    Returns
+    -------
+    threshold : float
+        Upper threshold value. All pixels with an intensity higher than
+        this value are assumed to be foreground.
+
+    References
+    ----------
+    .. [1] C. A. Glasbey, "An analysis of histogram-based thresholding
+        algorithms," CVGIP: Graphical Models and Image Processing,
+        vol. 55, pp. 532-537, 1993.
+
+    Examples
+    --------
+    >>> from skimage.data import camera
+    >>> image = camera()
+    >>> thresh = threshold_mean(image)
+    >>> binary = image > thresh
+    """
+    assert_nD(image, 2)
```

Comment:
Well, I think the solution to our regular issues could be to have a set of fused types like: ImageGrayscale2D, ImageGrayscale3D, ImageMultichannel3D, ImageGrayscaleND, ... or something more detailed to cover also never-ending issues with different dtype support and convertion: ImageFloat, ImageUint8, ...
The majority of our algorithms has limited support in terms of dimensionality and dtype (some support integer types, other support only floats). My personal opinion is that using this kind of a wrapper we could handle all of this cases more clearly. I'm aware if this discussion has already been carried out.
As an example, you might consider `pandas` with their `Dataframes` and fused dtypes for data columns, indices, etc (e.g. https://github.com/pydata/pandas/blob/master/pandas/types/dtypes.py).